A more pure design (#9)
oxinabox/LoggingExtras.jl: README.md

![Diagram showing how loggers connect](diag.svg)

# Discussion: Compositional Loggers

LoggingExtras is designs around allowing you to build arbitrarily complicated 
systems for "log plumbing". That is to say basically routing logged information to different places.
It is built around the idea of simple parts which are composed together,
to allow for powerful and flexible definition of your logging system.
Without having to define any custom loggers by subtyping `AbstractLogger`.
When we talk about composability we mean to say that the composition of any set of Loggers is itself a Logger.
LoggingExtras is a composable logging system.

Loggers can be broken down into 4 types:
 - *Sinks*: Sinks are the final end point of a log messages journey. They write it to file, or display it on the console, or set off a red flashing light in the laboratory. A Sink should never decide what to accept, only what to do with it.
 - *Filters*: Filters wrap around other loggers and decide wether or not to pass on a message. Thery can further be broken down by when that decision occurs (See `ActiveFilteredLogger` vs `EarlyFilteredLogger`).

>  - *Transformers*: Transformers modify the content of log messages, before passing them on. This includes the metadata like severity level. Unlike Filters they can't block a log message, but they could drop its level down to say `Debug` so that normally noone would see it.

Author: see https://github.com/oxinabox/LoggingExtras.jl/issues/10
which I think is an easy way to do this

>  - *Transformers*: Transformers modify the content of log messages, before passing them on. This includes the metadata like severity level. Unlike Filters they can't block a log message, but they could drop its level down to say `Debug` so that normally noone would see it.

Reviewer: But what type do you return from `MinLevelLogger(min_level)`? Is the wrapped sink set to `nothing`?

>  - *Transformers*: Transformers modify the content of log messages, before passing them on. This includes the metadata like severity level. Unlike Filters they can't block a log message, but they could drop its level down to say `Debug` so that normally noone would see it.

Author: Return a closure.
```
MinLevelLogger(min_level) = logger -> MinLevelLogger(logger, min_level)
```

So a partially setup logger is not a logger, just a closure,
but once setup with a logger, then their composition is a logger,

>  - *Transformers*: Transformers modify the content of log messages, before passing them on. This includes the metadata like severity level. Unlike Filters they can't block a log message, but they could drop its level down to say `Debug` so that normally noone would see it.

Reviewer: That's clever. Kind of bothers me that something which looks like a constructor isn't one but other than that...

>  - *Demux*: There is only one possible Demux Logger. and it is central to log routing. It acts as a hub that recieves 1 log message, and then sends copies of it to all its child loggers. Like iin the diagram above, it can be composed with Filters to control what goes where.

Author: see #1, I am leaning towards `Tee`

This is a basically full taxonomy of all compositional loggers.
Other than `Sinks`, this package implements the full set. So you shouldn't need to build your own routing components, just configure the ones included in this package.

It is worth understanding the idea of logging purity.
The loggers defined in this package are all pure.
The Filters, only filter, the Sinks only sink, the transformers only Transform.

We can contrast this to the the `ConsoleLogger` (the standard logger in the REPL).

> The `ConsoleLogger` is an in-pure sink.

Author: Or at least we need a `PureConsoleLogger`,
and a default logger for the REPL described in terms of the `PureConsoleLogger` and some operators.

As well as displaying logs to the user (as a Sink);
it uses the log content, in the form of the `max_log` kwarg to decide if a log should be displayed (Active Filtering);
and it has a min_enabled_level setting, that controls if it will accept a message at all 
(Early Filtering, in particular see `MinLevelLogger`).
If it was to be defined in a compositional way,
we would write;
```

> const ConsoleLogger(stream, min_level) = 

Author: It is not required.
I think I was thinking of 
`const ConsoleLogger = (stream, min_level) -> ...`

```suggestion
ConsoleLogger(stream, min_level) = 
```

# Loggers introduced by this package:

This package introduces 5 new loggers.
The `DemuxLogger`, the `FileLogger`, and 3 types of filtered logger.
All of them just wrap existing loggers.
 - The `DemuxLogger` sends the logs to multiple different loggers.

>  - The `FileLogger` is a simple logger sink that writes to file.

Author: ```suggestion
 - The `TransformerLogger` applies a function to modify log messages before passing them on.
 - The `FileLogger` is a simple logger sink that writes to file.
```

```

## `ActiveFilteredLogger`

The `ActiveFilteredLogger` exists to give more control over which messages should be logged.
It warps any logger, and before sending messages to the logger to log,
checks them against a filter function.
The filter function takes the full set of parameters of the message.
(See it's docstring with `?ActiveFilteredLogger` for more details.)

### Demo
We want to filter to only log strings staring with `"Yo Dawg!"`.

```
julia> function yodawg_filter(log_args)
	startswith(log_args.message, "Yo Dawg!")
end
 yodawg_filter (generic function with 1 method)                                                                                     

julia> filtered_logger = ActiveFilteredLogger(yodawg_filter, global_logger());

julia> with_logger(filtered_logger) do
	@info "Boring message"
	@warn "Yo Dawg! it is bad"
	@info "Another boring message"
	@info "Yo Dawg! it is all good"
end
┌ Warning: Yo Dawg! it is bad
└ @ Main REPL[28]:3
[ Info: Yo Dawg! it is all good
```

## `EarlyFilteredLogger`

The `EarlyFilteredLogger` is similar to the `ActiveFilteredLogger`,
but it runs earlier in the logging pipeline.
In particular it runs before the message is computed.

> It is (theoretically) useful to filter things early if creating the log message is expective.

Author: I've now seen Memento used *a lot*.
It doesn't support this and I am yet to hear complaints
but fair, will change to

```suggestion
It can be useful to filter things early if creating the log message is expensive.
```

> It is (theoretically) useful to filter things early if creating the log message is expective.

Reviewer: I dunno; I've just seen this actually happen in other systems which was why I tried to "design it away" in the logging frontend. Unfortunately it introduces some complexity and personally I don't feel completely happy with the way `shouldlog` and `min_enabled_level` turned out.

To explain some of the reasoning behind these functions:
* The reason for `min_enabled_level` was to allow for some very basic filtering without invoking dynamic dispatch at all becuase the `min_enabled_level` is cached in the logstate.
* On the other hand, `shouldlog` must invoke dynamic dispatch because the logger is loaded from the task local storage and can never be inferred. This is annoying.

So there's a choice here: which parts of early filtering go into some inflexible statically defined rules (currently only a level for all modules), which parts go into dynamically chosen rules (`shouldlog`) and which parts must be deferred until after message creation (`handle_message`)? If there's a way to integrate these more neatly that would be great.

           else
             return false
           end
         end
       end
make_throttled_logger (generic function with 1 method)

julia> throttled_logger = make_throttled_logger(Second(3));

julia> with_logger(throttled_logger) do
         for ii in 1:10
           sleep(1)
           @info "It happen" ii
         end
       end
┌ Info: It happen
└   ii = 1
┌ Info: It happen
└   ii = 4
┌ Info: It happen
└   ii = 7
┌ Info: It happen
└   ii = 10
```

## `MinLevelLogger`
This is basically a special case of the early filtered logger,
that just checks if the level of the message is above the level specified when it was created.

## `TransformerLogger`
The transformer logger allows for the modification of log messages.
This modification includes such things as its log level, and content,
and all the other arguments passed to `handle_message`.

When constructing a `TransformerLogger` you pass in a tranformation function,
and a logger to be wrapped.
The  transformation function takes a named tuple containing all the log message fields,
and should return a new modified named tuple.

A simple example of its use is truncating messages.

```
julia> using Logging, LoggingExtras

> julia> truncating_logger  = TransformerLogger(global_logger()) do log

Reviewer: We need to do this bundling of all the keys and values together, I like it.

One question - what happens if someone does
```
@info "asdf" message=123
```

I was wondering whether we should have a `LogRecord` type for this instead of a bare named tuple? The annoying thing is that it makes field access less convenient for the arbitrary key value pairs.

Reviewer: Oh never mind, I see that you end up with `log.message` and `log.kwargs.message` for this.  That makes sense and is the same as you'd get if we were to have a `LogRecord` type.

Author: Reason for a `NamedTuple` rather than a `struct` was so I could splat it back into the full form.
If we added a `LogRecord` struct to `Core.Logging` (and for backwards compat made the single args get wrapped into it),
Then I think that would be fine.

Reviewer: Yep I think we need to do something like that.

using Logging
using HTTP

function not_HTTP_message_filter(log)
	log._module != HTTP
end

> global_logger(EarlyFilteredLogger(not_HTTP_message_filter, global_logger()))

Reviewer: I've been thinking we should have a nice function to do this wrapping operation. This is a particular reason I'd like to be able to instantiate filters and transformers without talking about the sink.

```

## Raising HTTP debug level errors to be Info level

```
using LoggingExtras
using Logging
using HTTP

transformer_logger(global_logger()) do log
	if log._module == HTTP && log.level=Logging.Debug
		# Merge can be used to construct a new NamedTuple
		# which effectively is the overwriting of fields of a NamedTuple

> 		return merge(log, (; level=Logging.Info))

Reviewer: This is cool, and I think it's basically the right idea.

The only thing I don't like is that it looses the original context that the messages from HTTP were emitted as `@debug` level.  I think the problem is that `LogLevel` has two orthogonal concepts packed into one entity:

* A categorical concept: a broad "type of thing" that the event is
* A ordinal concept: the priority (ie, the verbose -> severe axis)

I feel that transformers should be encouraged to change the priority, but not the category. I think we can retrofit this somehow into the existing system without backward incompatibilities!

Author: difficult

Reviewer: Maybe not. I was thinking something like mapping

`LogLevel -> (category, priority)`

inside the first `handle_message(all_the_parameters)` while gathering `all_the_parameters` into a `LogRecord` or some such.

This also provides a cleaner interpretation of custom log levels which can then define their own version of this mapping rather than the current system where they define a conversion to `LogLevel` for the purposes of assessing priority.

Again, something to resolve in the stdlib rather than here.
